### PR TITLE
TagPaginate: Configurable 'page' keys #94

### PR DESCRIPTION
**Pull request by edwardoka**
- [x] I've run the tests with `vendor/bin/phpunit`
- [x] None of the tests were found failing
- [x] I've seen the coverage report at `build/coverage/index.html`
- [x] Not a single line left uncovered by tests
- [x] Any coding standards issues were fixed with `vendor/bin/php-cs-fixer fix`
- [x] Hotfix-type commits were squashed with `git rebase -i` or `git commit --amend`
- [x] All my work wasn't smushed into one large commit without necessity

### CODE REVIEW

File: src/Liquid/Tag/TagPaginate.php [hunk 1]
```diff
@@ -106,7 +106,11 @@ public function __construct($markup, array &$tokens, FileSystem $fileSystem = nu
 	 */
 	public function render(Context $context)
 	{
-		$this->currentPage = (is_numeric($context->get('page'))) ? $context->get('page') : 1;
+		// Check for a setting in the Liquid config for the context key to use to refer to the current page
+		$pageContextKey = Liquid::get('PAGINATION_CONTEXT_KEY');
+		$pageContextKey = ($pageContextKey) ? $pageContextKey : 'page';
```
Comment (sanmai) [inline src/Liquid/Tag/TagPaginate.php]: There's `Liquid::$config` where defaults are. Better have them in one place, what do you think?

File: src/Liquid/Tag/TagPaginate.php [hunk 2]
```diff
@@ -142,14 +146,18 @@ public function render(Context $context)
...
 		);
 
+		// look to see if there's a key in the Liquid config for the request field to use in URLs
+		$pageRequestKey = Liquid::get('PAGINATION_REQUEST_KEY');
+		$pageRequestKey = ($pageRequestKey) ? $pageRequestKey : 'page';
+
 		if ($this->currentPage != 1) {
 			$paginate['previous']['title'] = 'Previous';
-			$paginate['previous']['url'] = $this->currentUrl($context) . '?page=' . ($this->currentPage - 1);
+			$paginate['previous']['url'] = $this->currentUrl($context) . '?' . urlencode($pageRequestKey) . '=' . ($this->currentPage - 1);
```
Comment (sanmai) [inline src/Liquid/Tag/TagPaginate.php]: I would love to see here `http_build_query` used instead. Like so:

```php
http_build_query([
    $pageRequestKey => $this->currentPage - 1,
]);
```

File: tests/Liquid/Tag/TagPaginateTest.php [hunk 3]
```diff
@@ -67,4 +67,60 @@ public function testPaginationForRepeatedCalls()
...
+			'HTTP_HOST' => 'example.com', 'page' => 2, 'articles' => array(array('title' => 1), array('title' => 2), array('title' => 3))
+		);
+
+		$text = '{% paginate articles by 1 %}{% for article in articles %}{{article.title}}{% endfor %} {{paginate.previous.title}},{{paginate.previous.url}} {{paginate.next.title}},{{paginate.next.url}}{% endpaginate %}';
+
+		$expected = '2 Previous,http://example.com?pagina=1 Next,http://example.com?pagina=3';
+
+		\Liquid\Liquid::set('PAGINATION_REQUEST_KEY', 'pagina');
+		$this->assertTemplateResult($expected, $text, $assigns);
+		\Liquid\Liquid::set('PAGINATION_REQUEST_KEY', null);
```
Comment (sanmai) [inline tests/Liquid/Tag/TagPaginateTest.php]: This should save and restore the default value.

**Comment by sanmai** [comment]
Never mind the Codacy Review, it only complains about static methods. We can't fix them.

**Comment by edwardoka** [comment]
Excellent points, thanks. I'll push your requested changes shortly.

While looking at the code that generates next/previous URLs I happened to notice that it doesn't preserve any other query string parameters.

e.g. http://example.com/test.php?page=1&filter=test has all request params stripped by TagPaginate::currentUrl before ?page=2 is re-added.

I'm implementing a fix for this as well but will hold off pushing in case you want me to raise a new issue for this.

**Comment by edwardoka** [comment]
I've pushed my fix for the stripping of request parameters in this commit:
https://github.com/edwardoka/php-liquid/commit/c43dff0c254a92f6b1a2bdc826e2f5b46f66893d

I suspect I've made a mistake by taking this branch from my fix-94 branch instead of from master. Let me know if you need me to re-do this with a separate PR or if I'm ok to merge the changes in this commit back into fix-94.

**Comment by sanmai** [comment]
`currentUrl` is overly complicated already. Let me try to handle it.

**Comment by sanmai** [comment]
So far so good. Would you care to explain why do we need separate defaults for `PAGINATION_REQUEST_KEY` and `PAGINATION_CONTEXT_KEY`? It seemed like I understood what's that for at first, but now I lost it. And there's no comments in the code.

**Comment by edwardoka** [comment]
The reason for separate keys is that they are different concepts. While 'page' is a good default value for both, they should be able to be configured separately. 

In Asa's example in this comment https://github.com/kalimatas/php-liquid/issues/94#issuecomment-363748448], he uses the key 'page' in the array he passes to Template::render to pass data, presumably about the resource being visited - which conflicted the context key that TagPaginate uses to get its data.

I happened to notice while looking at the URL generation that the request parameter was hardcoded and it made sense to make this configurable as well, since we can't make assumptions about the user's application.

Thanks again for being patient with this newbie 👍

**Pull request merged**